exporter/signalfxexporter/dimensions/dimclient.go

```diff
@@ -0,0 +1,340 @@
+	// look up in case we need to replace due to flappy prop generation.
+	delayedSet map[DimensionKey]*DimensionUpdate
+	// Queue of dimensions to update.  The ordering should never change once
+	// put in the queue so no need for heap/priority queue.
+	delayedQueue chan *queuedDimension
+	// For easier unit testing
+	now func() time.Time
+
+	// TODO: Look into collecting these metrics and other traces via obsreport
+	DimensionsCurrentlyDelayed int64
+	TotalDimensionsDropped     int64
+	// The number of dimension updates that happened to the same dimension
+	// within sendDelay.
+	TotalFlappyUpdates           int64
+	TotalClientError4xxResponses int64
+	TotalRetriedUpdates          int64
+	TotalInvalidDimensions       int64
+	TotalSuccessfulUpdates       int64
+	logUpdates                   bool
+	logger                       *zap.Logger
+}
+
+type queuedDimension struct {
+	*DimensionUpdate
+	TimeToSend time.Time
+}
+
+type DimensionClientOptions struct {
+	Token      string
+	APIURL     *url.URL
+	LogUpdates bool
+	Logger     *zap.Logger
+	SendDelay  int
+}
+
+// NewDimensionClient returns a new client
+func NewDimensionClient(ctx context.Context, options DimensionClientOptions) *DimensionClient {
+
+	client := &http.Client{
+		Timeout: 10 * time.Second,
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   5 * time.Second,
+				KeepAlive: 30 * time.Second,
+				DualStack: true,
+			}).DialContext,
+			MaxIdleConns:        20,
+			MaxIdleConnsPerHost: 20,
+			IdleConnTimeout:     30 * time.Second,
+			TLSHandshakeTimeout: 10 * time.Second,
+		},
+	}
+	sender := NewReqSender(ctx, client, 20, map[string]string{"client": "dimension"})
+
+	return &DimensionClient{
+		ctx:           ctx,
+		Token:         options.Token,
+		APIURL:        options.APIURL,
+		sendDelay:     time.Duration(options.SendDelay) * time.Second,
+		delayedSet:    make(map[DimensionKey]*DimensionUpdate),
+		delayedQueue:  make(chan *queuedDimension, 10000),
+		requestSender: sender,
+		client:        client,
+		now:           time.Now,
+		logger:        options.Logger,
+		logUpdates:    options.LogUpdates,
+	}
+}
+
+// Start the client's processing queue
+func (dc *DimensionClient) Start() {
+	go dc.processQueue()
+}
+
+// acceptDimension to be sent to the API.  This will return fairly quickly and
+// won't block. If the buffer is full, the dim update will be dropped.
+func (dc *DimensionClient) acceptDimension(dimUpdate *DimensionUpdate) error {
+	dc.Lock()
+	defer dc.Unlock()
+
+	if delayedDimUpdate := dc.delayedSet[dimUpdate.Key()]; delayedDimUpdate != nil {
+		if !reflect.DeepEqual(delayedDimUpdate, dimUpdate) {
+			dc.TotalFlappyUpdates++
+		}
+
+		// Merge the latest updates into existing one.
+		delayedDimUpdate.Properties = mergeProperties(delayedDimUpdate.Properties, dimUpdate.Properties)
+		delayedDimUpdate.Tags = mergeTags(delayedDimUpdate.Tags, dimUpdate.Tags)
```

Comment:
Why do we need to do the merge in case if `delayedDimUpdate` and `dimUpdate`  are equal ? shouldn't be put under the condition above?